x/sync/manager.go

```diff
@@ -277,74 +347,55 @@ func (m *Manager) getAndApplyChangeProof(ctx context.Context, work *workItem) {
 		return
 	}
 
-	changeOrRangeProof, err := m.config.Client.GetChangeProof(
-		ctx,
-		&pb.SyncGetChangeProofRequest{
-			StartRootHash: work.localRootID[:],
-			EndRootHash:   targetRootID[:],
-			StartKey: &pb.MaybeBytes{
-				Value:     work.start.Value(),
-				IsNothing: work.start.IsNothing(),
-			},
-			EndKey: &pb.MaybeBytes{
-				Value:     work.end.Value(),
-				IsNothing: work.end.IsNothing(),
-			},
-			KeyLimit:   defaultRequestKeyLimit,
-			BytesLimit: defaultRequestByteSizeLimit,
+	request := &pb.SyncGetChangeProofRequest{
+		StartRootHash: work.localRootID[:],
+		EndRootHash:   targetRootID[:],
+		StartKey: &pb.MaybeBytes{
+			Value:     work.start.Value(),
+			IsNothing: work.start.IsNothing(),
 		},
-		m.config.DB,
-	)
+		EndKey: &pb.MaybeBytes{
+			Value:     work.end.Value(),
+			IsNothing: work.end.IsNothing(),
+		},
+		KeyLimit:   defaultRequestKeyLimit,
+		BytesLimit: defaultRequestByteSizeLimit,
+	}
+
+	requestBytes, err := proto.Marshal(request)
 	if err != nil {
+		m.finishWorkItem()
 		m.setError(err)
 		return
 	}
 
-	select {
-	case <-m.doneChan:
-		// If we're closed, don't apply the proof.
-		return
-	default:
-	}
+	onResponse := func(ctx context.Context, _ ids.NodeID, responseBytes []byte, err error) {
+		defer m.finishWorkItem()
 
-	if changeOrRangeProof.ChangeProof != nil {
-		// The server had sufficient history to respond with a change proof.
-		changeProof := changeOrRangeProof.ChangeProof
-		largestHandledKey := work.end
-		// if the proof wasn't empty, apply changes to the sync DB
-		if len(changeProof.KeyChanges) > 0 {
-			if err := m.config.DB.CommitChangeProof(ctx, changeProof); err != nil {
-				m.setError(err)
-				return
-			}
-			largestHandledKey = maybe.Some(changeProof.KeyChanges[len(changeProof.KeyChanges)-1].Key)
+		if err := m.handleChangeProofResponse(ctx, targetRootID, work, request, responseBytes, err); err != nil {
+			m.config.Log.Debug("dropping response", zap.Error(err))
```

Comment:
should we log some information about the request/response?

Reply:
logging the response will make this diff bigger and we don't already do it, but it's trivial to add logging the request.

Reply:
I think logging the request is good enough for now